torchelastic/distributed/launch.py

```diff
@@ -0,0 +1,369 @@
+assumptions about `WORLD_SIZE` as the world size can change due as
+nodes are allowed to leave and join.
+
+7. It is recommended your script have the following structure
+
+::
+    pre_initialize()
+    load_checkpoint(checkpoint_path)
+    initialize()
+    start_train()
+"""
+
+import os
+import subprocess
+import sys
+from argparse import REMAINDER, ArgumentParser
+
+import torch.distributed as dist
+import torchelastic.rendezvous.etcd_rendezvous  # noqa: F401
+from torchelastic.agent.server.api import WorkerSpec
+from torchelastic.agent.server.local_elastic_agent import LocalElasticAgent
+
+
+def parse_args(args):
+    """
+    Helper function parsing the command line options.
+    """
+
+    parser = ArgumentParser(description="torchelastic elastic training launcher")
+
+    # Arguments for the launch helper
+    # worker/node size related arguments
+    parser.add_argument(
+        "--nnodes",
+        type=str,
+        default="1:1",
+        help="number of nodes or MIN_NODES:MAX_NODES",
+    )
+    parser.add_argument(
+        "--nproc_per_node", type=int, default=1, help="number of workers per node"
+    )
+
+    # rendezvous related arguments
+    parser.add_argument(
+        "--rdzv_backend", type=str, default="etcd", help="rendezvous backend"
+    )
+    parser.add_argument(
+        "--rdzv_endpoint",
+        type=str,
+        default="",
+        help="rendezvous backend server host:port",
+    )
+    parser.add_argument("--rdzv_id", type=str, help="user defined group id")
+    parser.add_argument(
+        "--rdzv_conf",
+        type=str,
+        default="",
+        help="additional rdzv configuration (conf1=v1,conf2=v2,...)",
+    )
+
+    # user-code launch related arguments
+    parser.add_argument(
+        "--max_restarts",
+        type=int,
+        default=100,
+        help="max number of worker group restarts before failing",
+    )
+    parser.add_argument(
+        "--monitor_interval",
+        type=float,
+        default=60,
+        help="interval (in seconds) to monitor the state of workers",
+    )
+    parser.add_argument(
+        "--start_method",
+        type=str,
+        default="spawn",
+        choices=["spawn", "fork", "forkserver"],
+        help="multiprocessing start_method to use when creating workers",
+    )
+    parser.add_argument(
+        "--use_env",
+        default=False,
+        action="store_true",
+        help="Use environment variable to pass "
+        "'local rank'. For legacy reasons, the default value is False. "
+        "If set to True, the script will not pass "
+        "--local_rank as argument, and will instead set LOCAL_RANK.",
+    )
+
+    parser.add_argument(
+        "-m",
+        "--module",
+        default=False,
+        action="store_true",
+        help="Changes each process to interpret the launch script "
+        "as a python module, executing with the same behavior as"
+        "'python -m'.",
+    )
+    parser.add_argument(
+        "--no_python",
```

Comment:
if it is an executable, how do we resume from the previous checkpoint? Is it application developers' responsibility to make the executable recoverable? E.g., it should automatically save checkpoints, and when it is launched, it should always first look for previous checkpoints?

Reply:
yes it is up to the application to save, load, and keep track of checkpoints. In the future we will provide a mechanism for the application to call `torch.save(elastic-agent://path/checkpoint)` which will save the checkpoint with the agent, when the worker is restarted, it can retrieve the checkpoint locally from the agent without having to pull from a persistent store.